esp-hal-common/src/sha.rs

```diff
@@ -0,0 +1,517 @@
+            for (i, v) in to_write.chunks_exact(ALIGN_SIZE).enumerate() {
+                dst.add(i).write_volatile(u32::from_ne_bytes(v.try_into().unwrap()).to_be());
+            }
+        }
+
+        // If it's data we can't store we don't need to try and align it, just wait for next write
+        // Generally this applies when (src/4*4) != src
+        let was_bounded = dst_bound - to_write.len() == 0;
+        if remaining.len() > 0 && remaining.len() < 4 {
+            for i in 0..remaining.len() {
+                self.buf[i] = remaining[i];
+            }
+            self.buf_fill = remaining.len();
+            return (&[], was_bounded);
+        }
+
+        return (remaining, was_bounded);
+    }
+}
+
+
+#[derive(Debug)]
+pub struct Sha {
+    sha: SHA,
+    mode: ShaMode,
+    alignment_helper: AlignmentHelper,
+    cursor: usize,
+    first_run: bool,
+    finished: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ShaMode {
+    SHA1,
+    #[cfg(not(esp32))]
+    SHA224,
+    SHA256,
+    #[cfg(any(esp32s2, esp32s3, esp32))]
+    SHA384,
+    #[cfg(any(esp32s2, esp32s3, esp32))]
+    SHA512,
+    #[cfg(any(esp32s2, esp32s3))]
+    SHA512_224,
+    #[cfg(any(esp32s2, esp32s3))]
+    SHA512_256,
+    // SHA512_(u16) // Max 511
+}
+
+// TODO: Maybe make Sha Generic (Sha<Mode>) in order to allow for better
+// compiler optimizations? (Requires complex const generics which isn't stable
+// yet)
+
+#[cfg(esp32)]
+impl Sha {
+    pub fn new(sha: SHA, mode: ShaMode) -> Self {
+        // Setup SHA Mode
+        Self {
+            sha,
+            mode,
+            cursor: 0,
+            first_run: true,
+            finished: false,
+            alignment_helper: AlignmentHelper::default() 
+        }
+    }
+
+    fn process_buffer(&mut self) {
+        if self.first_run {
+            match self.mode {
+                ShaMode::SHA1 => self.sha.sha1_start().write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA256 => self.sha.sha256_start().write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA384 => self.sha.sha384_start.write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA512 => self.sha.sha512_start.write(|w| unsafe { w.bits(1u32) }),
+            }
+            self.first_run = false;
+        } else {
+            match self.mode {
+                ShaMode::SHA1 => self.sha.sha1_continue().write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA256 => self.sha.sha256_continue.write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA384 => self.sha.sha384_continue.write(|w| unsafe { w.bits(1u32) }),
+                ShaMode::SHA512 => self.sha.sha512_continue.write(|w| unsafe { w.bits(1u32) }),
+            }
+        }
+    }
+
+    fn is_busy(&mut self) -> bool {
+        match self.mode {
+            // FIXME: These are marked WO, while being RO
+            ShaMode::SHA1 => unsafe { self.sha.sha1_busy.as_ptr().read() == 0 },
```

Comment:
Found the problem! This is inverted, `0` is idle and `1` is processing data (aka busy). Flipping this makes the example work locally for me.